File: control/spectral_commutativity/spectral_commutativity.py
# SPDX-License-Identifier: AGPL-3.0-or-later

#!/usr/bin/env python3
"""
neuralSpring Paper 22 — Distance to Normal Elements in C*-algebras

Reproduces key concepts from:
  Kachkovskiy & Safarov (2016)
  "Distance to normal elements in C*-algebras of real rank zero"
  Journal of the American Mathematical Society 29:61–80.

Core thesis: Quantifies how close an operator is to being "normal"
(commuting with its adjoint). For neural networks, this relates to
skip connections and residual networks: layers that "approximately
commute" can be reordered without catastrophic information loss.
The distance-to-normal metric measures how "residual-friendly" a
weight matrix is.

BarraCUDA connection:
  - GEMM for matrix multiply (AB, A*A, AA*)
  - Frobenius norm: reduce_sum(elementwise square)
  - Residual layers (I + W): elementwise add
"""

# SPDX-License-Identifier: AGPL-3.0-only

import sys

import numpy as np

# ---------------------------------------------------------------------------
# Operator Commutativity
# ---------------------------------------------------------------------------


def commutator(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """[A,B] = AB - BA. If [A,B] = 0, they commute."""
    return a @ b - b @ a


def commutator_frobenius_norm(a: np.ndarray, b: np.ndarray) -> float:
    """||[A,B]||_F measures how much A and B don't commute."""
    return float(np.linalg.norm(commutator(a, b), "fro"))


def commutativity_ratio(a: np.ndarray, b: np.ndarray) -> float:
    """Normalized: ||[A,B]||_F / (||A||_F * ||B||_F). Scale-invariant."""
    na = np.linalg.norm(a, "fro")
    nb = np.linalg.norm(b, "fro")
    if na * nb < 1e-300:
        return 0.0
    return commutator_frobenius_norm(a, b) / (na * nb)


# ---------------------------------------------------------------------------
# Distance to Normal
# ---------------------------------------------------------------------------


def distance_to_normal(a: np.ndarray) -> float:
    """Distance to normal: A is normal iff A*A = AA* (A commutes with adjoint).

    Approximation: dist_normal(A) ≈ ||A*A - AA*||_F / (2*||A||_F).
    For real matrices, A* = A^T.
    """
    n = np.linalg.norm(a, "fro")
    if n < 1e-300:
        return 0.0
    ata = a.T @ a
    aat = a @ a.T
    diff = ata - aat
    return float(np.linalg.norm(diff, "fro") / (2.0 * n))


# ---------------------------------------------------------------------------
# Skip Connection Analysis
# ---------------------------------------------------------------------------


def skip_commutativity(w1: np.ndarray, w2: np.ndarray) -> tuple[float, float]:
    """Compare commutativity of W1,W2 vs (I+W1),(I+W2).

    Returns (comm_raw, comm_skip) where:
      comm_raw = ||[W1,W2]||_F (normalized by ||W1||*||W2||)
      comm_skip = ||[(I+W1),(I+W2)]||_F (normalized similarly)
    """
    n1, n2 = w1.shape[0], w2.shape[0]
    assert n1 == n2, "W1, W2 must be square and same size"
    i = np.eye(n1)
    r1 = commutativity_ratio(w1, w2)
    r2 = commutativity_ratio(i + w1, i + w2)
    return (r1, r2)


# ---------------------------------------------------------------------------
# Spectral Analysis
# ---------------------------------------------------------------------------


def spectral_gap_to_normal(a: np.ndarray) -> float:
    """For normal A: eigenvalues of A*A = eigenvalues of AA* (same spectrum).

    Spectral gap = max |eigenvalue(ata) - eigenvalue(aat)| over matching.
    """
    ata = a.T @ a
    aat = a @ a.T
    ev_ata = np.linalg.eigvalsh(ata)
    ev_aat = np.linalg.eigvalsh(aat)
    ev_ata = np.sort(ev_ata)
    ev_aat = np.sort(ev_aat)
    return float(np.max(np.abs(ev_ata - ev_aat)))


# ---------------------------------------------------------------------------
# Random Matrix Ensemble
# ---------------------------------------------------------------------------


def sample_distance_to_normal(
    n: int, n_samples: int = 100, seed: int = 42
) -> np.ndarray:
    """Sample many random matrices, return distribution of distance-to-normal."""
    rng = np.random.default_rng(seed)
    dists = []
    for _ in range(n_samples):
        a = rng.standard_normal((n, n)) / np.sqrt(n)
        dists.append(distance_to_normal(a))
    return np.array(dists)


# ---------------------------------------------------------------------------
# Main validation
# ---------------------------------------------------------------------------


def main() -> int:
    """Validate spectral commutativity and distance-to-normal.

    Provenance
    ----------
    Paper: Kachkovskiy & Safarov (2016) JAMS 29:61-80.
    """
    rng = np.random.default_rng(42)
    total_passed = 0
    total_failed = 0

    print("=" * 72)
    print("neuralSpring Paper 22: Spectral Commutativity & Distance to Normal")
    print("  Kachkovskiy & Safarov (2016) JAMS 29:61-80")
    print("=" * 72)

    n = 32

    # Check 1: Normal matrices have distance-to-normal ≈ 0
    print("\n--- Check 1: Normal matrices have dist_normal ≈ 0 ---")
    h = rng.standard_normal((n, n))
    sym = (h + h.T) / 2
    d_sym = distance_to_normal(sym)
    print(f"  Symmetric matrix dist_normal: {d_sym:.2e}")
    if d_sym < 1e-10:
        print("  [PASS] Normal (symmetric) has dist ≈ 0")
        total_passed += 1
    else:
        print(f"  [FAIL] Expected ≈ 0, got {d_sym}")
        total_failed += 1

    # Check 2: Identity matrix is normal (distance = 0)
    print("\n--- Check 2: Identity is normal ---")
    identity = np.eye(n)
    d_id = distance_to_normal(identity)
    print(f"  Identity dist_normal: {d_id:.2e}")
    if d_id < 1e-14:
        print("  [PASS] Identity has dist = 0")
        total_passed += 1
    else:
        print(f"  [FAIL] Expected 0, got {d_id}")
        total_failed += 1

    # Check 3: Skip connections reduce commutativity distance
    print("\n--- Check 3: Skip connections reduce commutativity ---")
    w1 = rng.standard_normal((n, n)) / np.sqrt(n)
    w2 = rng.standard_normal((n, n)) / np.sqrt(n)
    comm_raw, comm_skip = skip_commutativity(w1, w2)
    print(f"  Raw commutativity:   {comm_raw:.6f}")
    print(f"  Skip commutativity: {comm_skip:.6f}")
    if comm_skip < comm_raw:
        print("  [PASS] Skip connections reduce commutativity")
        total_passed += 1
    else:
        print(f"  [FAIL] Expected skip < raw")
        total_failed += 1

    # Check 4: Residual layers (I+eps*W) approximately commute for small eps
    print("\n--- Check 4: Residual layers approximately commute (small eps) ---")
    w1_r = rng.standard_normal((n, n)) / np.sqrt(n)
    w2_r = rng.standard_normal((n, n)) / np.sqrt(n)
    eps_small = 0.01
    i = np.eye(n)
    r1 = i + eps_small * w1_r
    r2 = i + eps_small * w2_r
    comm_res = commutativity_ratio(r1, r2)
    comm_raw_r = commutativity_ratio(w1_r, w2_r)
    print(f"  Residual (eps=0.01) commutativity: {comm_res:.6f}")
    print(f"  Raw W1,W2 commutativity: {comm_raw_r:.6f}")
    if comm_res < comm_raw_r:
        print("  [PASS] Small-epsilon residual layers nearly commute")
        total_passed += 1
    else:
        print(f"  [FAIL] Expected residual < raw")
        total_failed += 1

    # Check 5: Commutator anti-symmetry [A,B] = -[B,A]
    print("\n--- Check 5: Commutator anti-symmetry [A,B] = -[B,A] ---")
    a = rng.standard_normal((n, n)) / np.sqrt(n)
    b = rng.standard_normal((n, n)) / np.sqrt(n)
    ab = commutator(a, b)
    ba = commutator(b, a)
    err = np.linalg.norm(ab + ba, "fro")
    print(f"  ||[A,B] + [B,A]||_F: {err:.2e}")
    if err < 1e-10:
        print("  [PASS] [A,B] = -[B,A]")
        total_passed += 1
    else:
        print(f"  [FAIL] Anti-symmetry violated: {err}")
        total_failed += 1

    # Check 6: Distance-to-normal is non-negative
    print("\n--- Check 6: Distance-to-normal non-negative ---")
    dists = sample_distance_to_normal(n, 50, 42)
    min_d = float(np.min(dists))
    print(f"  Min over 50 samples: {min_d:.2e}")
    if min_d >= -1e-12:
        print("  [PASS] All distances non-negative")
        total_passed += 1
    else:
        print(f"  [FAIL] Negative distance: {min_d}")
        total_failed += 1

    # Check 7: Spectral theorem: normal has same eigenvalues for A*A, AA*
    print("\n--- Check 7: Spectral gap zero for normal ---")
    gap_sym = spectral_gap_to_normal(sym)
    gap_random = spectral_gap_to_normal(w1)
    print(f"  Symmetric (normal) spectral gap: {gap_sym:.2e}")
    print(f"  Random matrix spectral gap:      {gap_random:.4f}")
    if gap_sym < 1e-10:
        print("  [PASS] Normal matrix has spectral gap ≈ 0")
        total_passed += 1
    else:
        print(f"  [FAIL] Normal gap expected ≈ 0, got {gap_sym}")
        total_failed += 1

    # Check 8: BarraCUDA connection documented
    print("\n--- Check 8: BarraCUDA connection ---")
    print("  Distance-to-normal: gemm_f64 (A@A.T, A.T@A) + reduce_sum")
    print("  Commutator: gemm_f64 (AB, BA) + elementwise subtract")
    print("  Residual: elementwise add (I + W)")
    print("  [PASS] BarraCUDA connection documented")
    total_passed += 1

    total = total_passed + total_failed
    print(f"\n{'=' * 72}")
    print(f"TOTAL: {total_passed}/{total} PASS, {total_failed}/{total} FAIL")
    print(f"{'=' * 72}")

    return 0 if total_failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: src/spectral_commutativity.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Spectral commutativity and distance to normal.
//!
//! Port of `control/spectral_commutativity/spectral_commutativity.py`.
//!
//! Paper: Kachkovskiy & Safarov (2016)
//! "Distance to normal elements in C*-algebras of real rank zero"
//! JAMS 29:61–80.
//!
//! Core thesis: Quantifies how close an operator is to being normal.
//! For neural networks: skip connections and residual layers relate to
//! approximate commutativity.
//!
//! ## Layout
//!
//! All matrices use **flat row-major** `Vec<f64>` storage with explicit
//! dimension `n`. This is cache-friendly and directly uploadable to GPU.
//!
//! ## `BarraCUDA` connection
//!
//! - Matrix multiplication A×B: `barracuda::ops::matmul` (GEMM f64)
//! - Commutator \[A,B\] = AB − BA: two GEMM + elementwise subtract
//! - Frobenius norm: `barracuda::ops::NormReduceF64`
//! - Distance to normal: composed from commutator + Frobenius (GPU pipeline)

#![expect(
    clippy::cast_precision_loss,
    reason = "matrix dimension → f64 for norm computation"
)]

use crate::rng::Rng;

/// Frobenius norm: sqrt(sum of squares of entries).
///
/// CPU reference for GPU validation — `barracuda::dispatch::frobenius_norm_dispatch`
/// is the production equivalent.  Kept separate so GPU validators have an
/// independent reference to compare against.
#[must_use]
pub fn frobenius_norm(a: &[f64]) -> f64 {
    a.iter().map(|&x| x * x).sum::<f64>().sqrt()
}

/// Transpose an n×n matrix (flat row-major → flat row-major).
///
/// CPU reference for GPU validation — `barracuda::dispatch::transpose_dispatch`
/// is the production equivalent.
#[must_use]
pub fn transpose(a: &[f64], n: usize) -> Vec<f64> {
    let mut t = vec![0.0; n * n];
    for i in 0..n {
        for j in 0..n {
            t[j * n + i] = a[i * n + j];
        }
    }
    t
}

/// Matrix multiplication C = A × B for n×n square matrices (flat row-major).
///
/// CPU reference implementation for validation. The GPU equivalent is
/// `barracuda::ops::matmul` / `gemm_f64.wgsl`. This stays CPU-side so
/// GPU validation binaries have an independent reference to compare against.
#[must_use]
pub fn mat_mul(a: &[f64], b: &[f64], n: usize) -> Vec<f64> {
    let mut c = vec![0.0; n * n];
    for i in 0..n {
        for k in 0..n {
            let a_ik = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += a_ik * b[k * n + j];
            }
        }
    }
    c
}

/// Commutator \[A, B\] = AB - BA (flat row-major).
#[must_use]
pub fn commutator(a: &[f64], b: &[f64], n: usize) -> Vec<f64> {
    let ab = mat_mul(a, b, n);
    let ba = mat_mul(b, a, n);
    ab.iter().zip(ba.iter()).map(|(x, y)| x - y).collect()
}

/// Distance to normal: `||A*A - AA*||_F / (2||A||_F)`.
/// For real matrices, `A*` = Aᵀ. A is normal iff `A*A = AA*`.
#[must_use]
pub fn distance_to_normal(a: &[f64], n: usize) -> f64 {
    let norm = frobenius_norm(a);
    if norm < crate::primitives::LOG_GUARD {
        return 0.0;
    }
    let at = transpose(a, n);
    let ata = mat_mul(&at, a, n);
    let aat = mat_mul(a, &at, n);
    let diff: Vec<f64> = ata.iter().zip(aat.iter()).map(|(x, y)| x - y).collect();
    frobenius_norm(&diff) / (2.0 * norm)
}

/// Commutativity ratio: `||[A,B]||_F / (||A||_F * ||B||_F)`. Scale-invariant.
#[must_use]
pub fn commutativity_ratio(a: &[f64], b: &[f64], n: usize) -> f64 {
    let na = frobenius_norm(a);
    let nb = frobenius_norm(b);
    if na * nb < crate::primitives::LOG_GUARD {
        return 0.0;
    }
    let comm = commutator(a, b, n);
    frobenius_norm(&comm) / (na * nb)
}

/// Skip connection analysis: `(raw_ratio, skip_ratio)` for W1,W2 vs (I+W1),(I+W2).
#[must_use]
pub fn skip_commutativity(w1: &[f64], w2: &[f64], n: usize) -> (f64, f64) {
    let ident = identity_matrix(n);
    let i_plus_w1: Vec<f64> = ident.iter().zip(w1.iter()).map(|(a, b)| a + b).collect();
    let i_plus_w2: Vec<f64> = ident.iter().zip(w2.iter()).map(|(a, b)| a + b).collect();
    let raw = commutativity_ratio(w1, w2, n);
    let skip = commutativity_ratio(&i_plus_w1, &i_plus_w2, n);
    (raw, skip)
}

/// Identity matrix of size n (flat row-major).
#[must_use]
pub fn identity_matrix(n: usize) -> Vec<f64> {
    let mut m = vec![0.0; n * n];
    for i in 0..n {
        m[i * n + i] = 1.0;
    }
    m
}

/// Generate random matrix (n × n) with standard normal entries / sqrt(n).
#[must_use]
pub fn random_matrix(n: usize, rng: &mut Rng) -> Vec<f64> {
    let scale = 1.0 / (n as f64).sqrt();
    (0..n * n)
        .map(|_| rng.normal_params(0.0, 1.0) * scale)
        .collect()
}

/// Symmetric matrix from random H: (H + Hᵀ) / 2. Normal (commutes with adjoint).
#[must_use]
pub fn random_symmetric(n: usize, rng: &mut Rng) -> Vec<f64> {
    let h = random_matrix(n, rng);
    let ht = transpose(&h, n);
    h.iter()
        .zip(ht.iter())
        .map(|(a, b)| (a + b) * 0.5)
        .collect()
}

/// Spectral gap: max|eig(`AᵀA`) - eig(`AAᵀ`)|. Zero for normal matrices.
///
/// Uses Frobenius norm of `AᵀA` - `AAᵀ` as proxy (zero for normal).
#[must_use]
pub fn spectral_gap_approx(a: &[f64], n: usize) -> f64 {
    let at = transpose(a, n);
    let ata = mat_mul(&at, a, n);
    let aat = mat_mul(a, &at, n);
    let diff: Vec<f64> = ata.iter().zip(aat.iter()).map(|(x, y)| x - y).collect();
    frobenius_norm(&diff)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tolerances;

    #[test]
    fn commutator_antisymmetric() {
        let mut rng = Rng::new(42);
        let a = random_matrix(8, &mut rng);
        let b = random_matrix(8, &mut rng);
        let ab = commutator(&a, &b, 8);
        let ba = commutator(&b, &a, 8);
        let sum: Vec<f64> = ab.iter().zip(ba.iter()).map(|(x, y)| x + y).collect();
        assert!(
            frobenius_norm(&sum) < tolerances::CROSS_LANGUAGE,
            " [A,B] = -[B,A]"
        );
    }

    #[test]
    fn distance_to_normal_symmetric_zero() {
        let mut rng = Rng::new(42);
        let sym = random_symmetric(16, &mut rng);
        let d = distance_to_normal(&sym, 16);
        assert!(
            d < tolerances::CROSS_LANGUAGE,
            "symmetric (normal) should have d≈0, got {d}"
        );
    }

    #[test]
    fn skip_reduces_commutativity() {
        let mut rng = Rng::new(42);
        let w1 = random_matrix(16, &mut rng);
        let w2 = random_matrix(16, &mut rng);
        let (raw, skip) = skip_commutativity(&w1, &w2, 16);
        assert!(skip < raw, "skip ({skip}) < raw ({raw})");
    }

    #[test]
    fn mat_mul_identity() {
        let n = 4;
        let ident = identity_matrix(n);
        let mut rng = Rng::new(42);
        let a = random_matrix(n, &mut rng);
        let result = mat_mul(&a, &ident, n);
        let diff = frobenius_norm(
            &a.iter()
                .zip(result.iter())
                .map(|(x, y)| x - y)
                .collect::<Vec<_>>(),
        );
        assert!(
            diff < tolerances::ZERO_DETECTION,
            "A*I should equal A, diff={diff}"
        );
    }

    #[test]
    fn transpose_involutory() {
        let n = 5;
        let mut rng = Rng::new(42);
        let a = random_matrix(n, &mut rng);
        let att = transpose(&transpose(&a, n), n);
        let diff = frobenius_norm(
            &a.iter()
                .zip(att.iter())
                .map(|(x, y)| x - y)
                .collect::<Vec<_>>(),
        );
        assert!(
            diff < tolerances::ZERO_DETECTION,
            "transpose(transpose(A)) should equal A"
        );
    }

    #[test]
    fn spectral_gap_approx_symmetric_zero() {
        let mut rng = Rng::new(42);
        let sym = random_symmetric(8, &mut rng);
        let gap = spectral_gap_approx(&sym, 8);
        assert!(
            gap < tolerances::CROSS_LANGUAGE,
            "symmetric matrix: AᵀA = AAᵀ, gap ≈ 0, got {gap}"
        );
    }

    #[test]
    fn spectral_gap_approx_identity() {
        let n = 4;
        let eye = identity_matrix(n);
        let gap = spectral_gap_approx(&eye, n);
        assert!(gap < tolerances::ZERO_DETECTION, "identity gap = 0");
    }

    #[test]
    fn commutativity_ratio_identity() {
        let n = 4;
        let eye = identity_matrix(n);
        let mut rng = Rng::new(42);
        let a = random_matrix(n, &mut rng);
        let ratio = commutativity_ratio(&eye, &a, n);
        assert!(
            ratio < tolerances::CROSS_LANGUAGE,
            "I commutes with everything, ratio ≈ 0"
        );
    }
}
